### PR TITLE
Trainer card

### PR DESCRIPTION
**Pull request by ghoulslash**
decompile trainer_card.s

### CODE REVIEW

File: src/trainer_card.c [hunk 1]
```diff
@@ -0,0 +1,1958 @@
...
+}
+
+static void TrainerCard_PrintPokemonIconsOnCard(void)
+{
+    u8 i;
+    u8 buffer[6];
+    u8 buffer2[6];
+
+    memcpy(buffer, sUnknown_83CD94C, 6);
+    memcpy(buffer2, sUnknown_83CD952, 6);
```
Comment (jiangzhengwenjz) [inline src/trainer_card.c]: sizeof

File: src/trainer_card.c [hunk 2]
```diff
@@ -0,0 +1,1958 @@
...
+        LoadBgTiles(3, GetMonIconTiles(sTrainerCardDataPtr->trainerCard.monSpecies[i], 0), 512, 16 * i + 32);
+    }
+}
+
+static void sub_808ABE0(void)
+{
+    u8 i;
+    u8 buffer[4];
+    
+    memcpy(buffer, sUnknown_83CD958, 4);
```
Comment (jiangzhengwenjz) [inline src/trainer_card.c]: sizeof

File: src/trainer_card.c [hunk 3]
```diff
@@ -0,0 +1,1958 @@
...
+        FillBgTilemapBufferRect_Palette0(2, 0, 0, 0, 32, 32);
+        FillBgTilemapBufferRect_Palette0(3, 0, 0, 0, 32, 32);
+        return TRUE;
+    }
+    
+    sTrainerCardDataPtr->bgPalLoadState++;
+    return FALSE;
+}
+
+static void LoadTrainerCardTilemap2(u16* ptr)
```
Comment (jiangzhengwenjz) [inline src/trainer_card.c]: add a const qualifier to argument

File: src/trainer_card.c [hunk 4]
```diff
@@ -0,0 +1,1958 @@
...
+                dst[32 * i + j] = ptr[30 * i + j];
+            else
+                dst[32 * i + j] = ptr[0];
+        }
+    }
+    
+    CopyBgTilemapBufferToVram(2);
+}
+
+static void LoadTrainerCardTilemap0(u16* ptr)
```
Comment (jiangzhengwenjz) [inline src/trainer_card.c]: add a const qualifier to argument

File: src/trainer_card.c [hunk 5]
```diff
@@ -0,0 +1,1958 @@
...
+    CopyBgTilemapBufferToVram(3);
+}
+
+static void sub_808B180(void)
+{
+    if (++sTrainerCardDataPtr->var_6 > 60)
+    {
+        sTrainerCardDataPtr->var_6 = 0;
+        sTrainerCardDataPtr->var_7 ^= 1;
+        sTrainerCardDataPtr->var_7BCE = 1;
```
Comment (jiangzhengwenjz) [inline src/trainer_card.c]: `TRUE`

File: src/trainer_card.c [hunk 6]
```diff
@@ -0,0 +1,1958 @@
...
+    }
+}
+
+u8 GetTrainerCardStars(u8 cardId)
+{
+    struct TrainerCard* trainerCards = gTrainerCards;
```
Comment (jiangzhengwenjz) [inline src/trainer_card.c]: Are you sure you need this temp var?

File: src/trainer_card.c [hunk 7]
```diff
@@ -0,0 +1,1958 @@
...
+    var_24 += r5 * r4;
+    r10 = r5 / r4;
+    r5 *= 2;
+
+    for (i = 0; i < r7; i++)
+    {
+        gScanlineEffectRegBuffers[0][i] = -i;
+    }
+    
+    for (; i < (s16)(r9); i++)
```
Comment (jiangzhengwenjz) [inline src/trainer_card.c]: `i < (s16)r9`

File: src/trainer_card.c [hunk 8]
```diff
@@ -0,0 +1,1958 @@
...
+        if (sTrainerCardDataPtr->cardType != CARD_TYPE_FRLG)
+        {
+            sub_810C330(FacilityClassToPicIndex(val), TRUE, sTrainerCardPicPositions[sTrainerCardDataPtr->cardType][sTrainerCardDataPtr->trainerCard.gender][0],
+                    sTrainerCardPicPositions[sTrainerCardDataPtr->cardType][sTrainerCardDataPtr->trainerCard.gender][1], 8, 2);
+        }
+        else
+        {
+            sub_810C330(PlayerGenderToFrontTrainerPicId_Debug(sTrainerCardDataPtr->trainerCard.gender, TRUE), TRUE,
+                    sTrainerCardPicPositions[sTrainerCardDataPtr->cardType][sTrainerCardDataPtr->trainerCard.gender][0],
+                    sTrainerCardPicPositions[sTrainerCardDataPtr->cardType][sTrainerCardDataPtr->trainerCard.gender][1],
```
Comment (jiangzhengwenjz) [inline src/trainer_card.c]: sizeof

Comment (jiangzhengwenjz) [inline src/trainer_card.c]: sizeof

File: src/trainer_card.c [hunk 9]
```diff
@@ -0,0 +1,1958 @@
...
+    /*0x000A*/ bool8 hasPokedex;
+    /*0x000B*/ bool8 hasHofResult;
+    /*0x000C*/ bool8 hasLinkResults;
+    /*0x000D*/ bool8 hasBattleTowerWins;
+    /*0x000E*/ u8 var_E;
+    /*0x000F*/ u8 var_F;
+    /*0x0010*/ bool8 hasTrades;
+    /*0x0011*/ u8 hasBadge[BADGE_COUNT];
+    /*0x0019*/ u8 var_19[4][0xD];
+    /*0x004D*/ u8 var_4D[0x46];
```
[the remark this answers is not in the document]
Comment (ghoulslash) [reply src/trainer_card.c]: you mean all of them (eg. var_4D[12][0x46])? or this element specifically

[the remark this answers is not in the document]
Comment (jiangzhengwenjz) [reply src/trainer_card.c]: all of them

File: src/trainer_card.c [hunk 10]
```diff
@@ -0,0 +1,1958 @@
...
+    .linkBattleLosses = 5535,
+    .battleTowerWins = 65535,
+    .battleTowerStraightWins = 65535,
+    .contestsWithFriends = 55555,
+    .pokeblocksWithFriends = 44444,
+    .pokemonTrades = 33333    ,
+    .money = 999999,
+    .var_28 = {0, 0, 0, 0},
+    .playerName = _("るびさふぁこ！"),
+    .version = 0,
```
[the remark this answers is not in the document]
Comment (ghoulslash) [reply src/trainer_card.c]: they start with VERSION_SAPPHIRE = 1. I can add VERSION_LINK or UNK?

[the remark this answers is not in the document]
Comment (jiangzhengwenjz) [reply src/trainer_card.c]: Sorry, it's my fault. Forget about this one.

File: src/trainer_card.c [hunk 11]
```diff
@@ -59,61 +77,16 @@ struct TrainerCardData
...
-    /*0x01AB*/ u8 var_1AB[0x46];
-    /*0x01F1*/ u8 var_1F1[0x46];
-    /*0x0237*/ u8 var_237[0x46];
-    /*0x027D*/ u8 var_27D[0x46];
-    /*0x02C3*/ u8 var_2C3[0x46];
-    /*0x0309*/ u8 var_309[0x46];
-    /*0x034F*/ u8 var_34F[0x46];
+    /*0x0011*/ bool8 hasBadge[BADGE_COUNT];
+    /*0x0019*/ u8 var_19[4][13];
+    /*0x004D*/ u8 strings[12][70];
```
Comment (jiangzhengwenjz) [inline src/trainer_card.c]: add a `TRAINER_CARD_STRING_COUNT` entry to the enum above, and replace 12 with it

**Pull request merged**